### ISSUE TITLE
alertmanager failed to get alert, alert not found

### ISSUE THREAD

**Issue by karancode** [coreos/kube-prometheus#205]
**What happened?**
We've deployed kube-prometheus to our eks cluster and have configured alertmanager to send alert notification to slack.

Alertmanager is not sending the alert notification(in this case, specifically for the resolved notification) for some of the triggered alerts.

**Did you expect to see some different?**
Alertmanager should've send the notitication for all the triggered alerts.

**How to reproduce it (as minimally and precisely as possible)**:
Deployed kube-prometheus stock to our eks cluster using `kubectl apply`

**Environment**
_EKS_ 
Version `1.13`
_Kube-prometheus_
Version  `2.7.2`
Revision `82f98c825a1437c0a24056eb4dc986d9a3fc7fae`
Branch `HEAD`

* Prometheus Operator version:

    ```quay.io/coreos/prometheus-operator:v0.29.0```

* Kubernetes version information:

```
$ kubectl version --short
Client Version: v1.13.0
Server Version: v1.13.8-eks-a977ba
```

* Kubernetes cluster kind:

    we are using `Terraform` to setup our eks.

* Manifests:

Alertmanager config : 
```
global:
  resolve_timeout: 5m

route:
  group_by: ['job', 'alertname', 'severity', 'level']
  group_wait: 30s
  group_interval: 5s
  repeat_interval: 3h
  receiver: alert-channel

  routes:
  - match: 
      severity: critical
    receiver: alert-channel

  - match:
      severity: warning
    receiver: alert-channel-2

  - match:
  # watchdog alert
      alertname: Watchdog
    repeat_interval: 3h
    receiver: alert-channel-2

receivers: 
- name: 'alert-channel'
  slack_configs:
  - api_url: xxxx
    channel: 'xxx'
    title: '[{{ .Status | toUpper }}{{ if eq .Status "firing" }}:{{ .Alerts.Firing | len }}{{ end }}] [Prometheus Event]
    Alert'
    text: |-
      {{ range .Alerts }}
        {{- if .Annotations.summary }}*Alert:* {{ .Annotations.summary }} - `{{ .Labels.level }}`{{- end }}
        *Description:* {{ .Annotations.description }}{{ .Annotations.message }}
        *Details:*
        {{ range .Labels.SortedPairs }} • *{{ .Name }}:* `{{ .Value }}`
        {{ end }}
      {{ end }}
    send_resolved: true
  - api_url: yyyyyy
    channel: 'yyyyy'     
    title: '[{{ .Status | toUpper }}{{ if eq .Status "firing" }}:{{ .Alerts.Firing | len }}{{ end }}] 
    Alert'
    text: |-
      {{ range .Alerts }}
        {{- if .Annotations.summary }}*Alert:* {{ .Annotations.summary }} - `{{ .Labels.level }}`{{- end }}
        *Description:* {{ .Annotations.description }}{{ .Annotations.message }}
        *Details:*
        {{ range .Labels.SortedPairs }} • *{{ .Name }}:* `{{ .Value }}`
        {{ end }}
      {{ end }}
    send_resolved: true
```

* Prometheus Logs:

I see many such logs from the `prometheus-k8s-1` pod.
```
prometheus-k8s-1 prometheus level=warn ts=2019-09-03T04:00:31.853885596Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/app/discovery/kubernetes/kubernetes.go:300: watch of *v1.Endpoints ended with: too old resource version: 7582599 (7584913)"
prometheus-k8s-1 prometheus level=warn ts=2019-09-03T04:06:33.12787459Z caller=klog.go:86 component=k8s_client_runtime func=Warningf msg="/app/discovery/kubernetes/kubernetes.go:300: watch of *v1.Endpoints ended with: too old resource version: 7585076 (7587229)"
```

* Alertmanager Logs:

From the alertmanager pods I could see these logs : 
```
alertmanager-main-2 alertmanager level=error ts=2019-09-02T08:10:05.194932139Z caller=dispatch.go:354 component=dispatcher aggrGroup="{}/{severity=\"critical\"}:{alertname=\"KubeStatefulSetReplicasMismatch\", job=\"kube-state-metrics\", severity=\"critical\"}" msg="failed to get alert" err="alert not found"
alertmanager-main-2 alertmanager level=error ts=2019-09-02T08:48:35.350447361Z caller=dispatch.go:354 component=dispatcher aggrGroup="{}/{severity=\"critical\"}:{alertname=\"KubeStatefulSetReplicasMismatch\", job=\"kube-state-metrics\", severity=\"critical\"}" msg="failed to get alert" err="alert not found"
alertmanager-main-2 alertmanager level=error ts=2019-09-02T09:55:05.243382586Z caller=dispatch.go:354 component=dispatcher aggrGroup="{}/{severity=\"critical\"}:{alertname=\"KubeStatefulSetUpdateNotRolledOut\", job=\"kube-state-metrics\", severity=\"critical\"}" msg="failed to get alert" err="alert not found"
```
Not sure how to detect these events or how to resolve this.

**Comment by brancz** [comment]
cc @simonpasquier

**Comment by simonpasquier** [comment]
Which version of AlertManager please?

**Comment by karancode** [comment]
@simonpasquier it's
```
Branch: HEAD
GoVersion: go1.11.5
Revision: 571caec278be1f0dbadfdf5effd0bbea16562cfc
Version: 0.16.1
```

**Comment by hawran** [comment]
Hi,
we've just experienced the same problem:  alertmanager has not sent the resolved notification.

I haven't found any clue to possible culprits, neither within logs nor in alertmanager's metrics.
However, what has struck my eye was a bunch of error messages within alertmanager's logs as follows:
```
...
level=error ts=2019-09-16T17:49:13.779Z caller=dispatch.go:443 component=dispatcher aggrGroup=... msg="failed to get alert" err="alert not found"
level=error ts=2019-09-17T05:19:13.788Z caller=dispatch.go:443 component=dispatcher aggrGroup=... msg="failed to get alert" err="alert not found"
level=error ts=2019-09-17T07:20:13.765Z caller=dispatch.go:443 component=dispatcher aggrGroup=... msg="failed to get alert" err="alert not found"
level=error ts=2019-09-17T07:20:13.765Z caller=dispatch.go:443 component=dispatcher aggrGroup=... msg="failed to get alert" err="alert not found"
level=error ts=2019-09-17T08:48:13.794Z caller=dispatch.go:443 component=dispatcher aggrGroup=... msg="failed to get alert" err="alert not found"
...
```
Regarding the comment above the line https://github.com/prometheus/alertmanager/blob/master/dispatch/dispatch.go#L443 one must ask what's going on here?

Alermanager:
```
GoVersion: go1.11.12
Revision: v0.19.0
Version: 0.19.0
```
Prometheus:
```
GoVersion: go1.12.9
Version: 2.12.0
```

**Comment by simonpasquier** [comment]
After looking at the AlertManager code more thoroughly, I think that I've identified a race where resolved alerts could be garbage collected before being notified to the receiver(s). I'll continue my investigation and post updates here.

**Comment by simonpasquier** [comment]
See https://github.com/prometheus/alertmanager/pull/2040

**Comment by hawran** [comment]
Hi, any progress on the matter?
(a new release ideally, ;-))

**Comment by brancz** [comment]
I don't think this has been released yet but @simonpasquier would know best.

**Comment by simonpasquier** [comment]
Not yet released but we'll probably do a new version soon.

**Comment by hawran** [comment]
OK, thank you.

**Comment by skokkira** [comment]
I am encountering same issue ..any body found any thing

**Comment by simonpasquier** [comment]
Closing this issue since the fix has been included in [Alertmanager v0.20.0](https://github.com/prometheus/alertmanager/releases/tag/v0.20.0).

**Comment by AndrewTheBoss1** [comment]
NO DON"T CLOSE IT"S BROKEN!
